Add initial environments for Young and Michael

giordano opened the pull request:
Ref: https://github.com/UCL-ARC/rhel9-migration/issues/81.  Not sure I got everything right, but this is a start.  I'm also not sure about where to put the environment files, I put them in the same directory as the Kathleen one for now.

giordano committed:
Add initial environments for Young and Michael

giordano commented on spacksites/spack-env-templates/slurm/avx2.yaml:
```diff
@@ -0,0 +1,26 @@
+    - - '%gcc@12.3.0'
+
+  include:
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_first.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu-on-gpu.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_third.yaml
```
I didn't find the resolved `gpu_first` & Co. environments in `/lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/` to be included with `include_concrete`, so I put these ones for now.

heatherkellyucl replied:
Yes, we need those environments' .yaml files to be put in spack-env-templates/slurm (and none of them have been built in dev-b so far).

giordano replied:
I copied the gpu packages to this directory, so now I can do
```suggestion
  include_concrete:
  - ${SPACK_ROOT}/var/spack/environments/gpu_first
  - ${SPACK_ROOT}/var/spack/environments/gpu-on-gpu
  - ${SPACK_ROOT}/var/spack/environments/gpu_third
```
like https://github.com/UCL-ARC/hpc-spack/blob/0e6890c961cfdaf3f9ecd6ff060dde23628a7ce8/spacksites/spack-env-templates/dev1/build/myriad.yaml#L10-L15
?

heatherkellyucl replied:
Yes.

giordano commented on spacksites/spack-env-templates/slurm/multiarch.yaml:
I put in here all the packages from `base` that weren't mentioned in https://github.com/UCL-ARC/rhel9-migration/issues/81#issue-3063329303 for `avx2.yaml`, but I'm not sure all these packages are good for this, like gsl, castep, gromacs, cp2k and lammps

heatherkellyucl replied:
multiarch shouldn't have any of the end-use applications (gsl, castep, gromacs, cp2k and lammps) or any libraries that aren't multi-kernel/multiarch - those need to go in avx2 and avx512 and be built twice. I think it should only have the ones I listed in the issue, unless you know of others.

openblas making sure dynamic arch is on
fftw
python
[julia - not now]
intel
**openmpi** - I didn't mention that one.

and I think they should not be built for `x86_64_v3` (?) because they should use avx512 instructions if available at runtime and will fall back if it isn't (or not use them at all, like python and openmpi). I'm not entirely sure how that works with the target arch being set.

giordano replied:
Done!  But to be clear, is "intel" the Intel compiler (which is already there), or the Intel packages like `intel-oneapi-mkl`?

> and I think they should not be built for `x86_64_v3` (?) because they should use avx512 instructions if available at runtime and will fall back if it isn't (or not use them at all, like python and openmpi). I'm not entirely sure how that works with the target arch being set.

Note that  `x86_64_v3` is ~avx2, instead ~avx512 would be level 4 (which is what I used for the `avx512.yaml` environment): https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels.  However I really think these packages should be built for the common denominator (~avx2 -> level 3), because where they can do multiversioning they simply ignore the compiler optimisations (they either directly call the assembler on hand-written assembly kernels, or anyway their buildsystems force the necessary opt flags for the specialised C kernels), but when the compiler optimisations are relevant you want to target the lower baseline, otherwise you may get illegal instruction errors at runtime. Python doesn't do any multiversioning at all as far as I know, and the Julia executable definitely doesn't (but the code generated by the Julia compiler can [optionally be multiversioned](https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_CPU_TARGET))

heatherkellyucl replied:
Sorry, "intel" is all the intel packages, including mkl and mpi. The things from the intel environment that aren't vasp or hdf5 .

```
    - - intel-oneapi-mpi@2021.14.0 +external-libfabric
      # VASP is tested with 2023.2.0. +cluster gives scalapack
      - intel-oneapi-mkl@2023.2.0 +cluster mpi_family=mpich ^intel-oneapi-mpi
      - intel-oneapi-mkl@2023.2.0 +cluster mpi_family=openmpi ^openmpi%oneapi@2024.2.1
      - openmpi@4.1.6 fabrics=psm2,ucx,ofi schedulers=slurm +legacylaunchers
      - - '%oneapi@2024.2.1' # VASP is tested with 2024.2
  - intel-oneapi-compilers@2024.2.1
```

Hmm, the multiarch environment was supposed to be things that aren't build tools that only need to be built once, either because they do not use avx512 (python - that would come via openblas), or the idea was that they be built with multiple kernels and use avx512 or avx2 at runtime. 

> because where they can do multiversioning they simply ignore the compiler optimisations (they either directly call the assembler on hand-written assembly kernels, or anyway their buildsystems force the necessary opt flags for the specialised C kernels), but when the compiler optimisations are relevant you want to target the lower baseline, otherwise you may get illegal instruction errors at runtime.

Do you mean the compiler optimisations used for building them, or when a compiler builds things using them? (On reread I think the first - ok, I think that makes sense). Is there likely to be a performance impact on the avx512 nodes from not having those compiler optimisations for them, in that case?

giordano replied:
> Is there likely to be a performance impact on the avx512 nodes from not having those compiler optimisations for them, in that case?

No, for openblas (when `dynamic_dispatch` is enabled, which is the default but I also enabled it here not to rely on default values) and fftw there should be no performance impact at all: on whatever host you compile them, all the optimised kernels for the given architecture are automatically compiled and those library will automatically pick up the best kernels for the CPU used at runtime, detected via [CPUID](https://en.wikipedia.org/wiki/CPUID) or similar mechanisms.

heatherkellyucl commented:
There does also need one final environment that include_concretes all the others, to make the whole stack.

heatherkellyucl commented on spacksites/spack-env-templates/slurm/avx512.yaml:
This one shouldn't include the GPU environments unless this is the bottom/encompassing one of the whole stack.

It is the Young and Michael avx512 end applications.

heatherkellyucl replied:
(and should include all of castep, lammps, gromacs etc as already mentioned).

giordano commented on spacksites/spack-env-templates/slurm/avx512.yaml:
```diff
@@ -0,0 +1,26 @@
+  include:
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_first.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu-on-gpu.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_third.yaml
```
Should I remove the GPU packages from the avx512 environment?  Based on https://github.com/UCL-ARC/rhel9-migration/issues/81#issue-3063329303 we don't have GPU avx512 nodes, right?

heatherkellyucl replied:
Yes - the gpu nodes are all avx2.

giordano committed:
Move end applications back to avx2/avx512 environments

giordano committed:
Copy GPU environments to the slurm templates

giordano reviewed (commented):
~It's not entirely clear to me what I should do for the gpu environments.~ ***Edit***: maybe that'd be answered by https://github.com/UCL-ARC/hpc-spack/pull/68#discussion_r2109943562?

giordano committed:
Include new concrete GPU environments

heatherkellyucl commented on spacksites/spack-env-templates/slurm/avx2.yaml:
```diff
@@ -0,0 +1,45 @@
+      - samtools@1.17
+    - - '%gcc@12.3.0'
+
+  include_concrete:
+  - ${SPACK_ROOT}/var/spack/environments/gpu_first
+  - ${SPACK_ROOT}/var/spack/environments/gpu-on-gpu
+  - ${SPACK_ROOT}/var/spack/environments/gpu_third
```
I don't think using ${SPACK_ROOT} here works, which was why I had to go to full paths for the other environments.

Need to make sure that these get set wherever castep, orca and vasp packages get installed (I guess could go into the various `settings/initial_site_packages_` yamls now):

```
  packages:
    castep:
      permissions:
        read: group
        group: ag-archpc-castep
    orca:
      permissions:
        read: group
        group: ag-archpc-orca
    vasp@6:
      permissions:
        read: group
        group: ag-archpc-vasp6
```

giordano replied:
> I don't think using ${SPACK_ROOT} here works, which was why I had to go to full paths for the other environments.

Agreed, I [mentioned](https://github.com/UCL-ARC/hpc-spack/pull/68#discussion_r2104301091) this was mainly a placeholder 🙂 But it's still not clear to me what should be the absolute paths.

> Need to make sure that these get set wherever castep, orca and vasp packages get installed 

Done!

heatherkellyucl replied:
The paths for the Slurm environments should be `/lustre/apps/spack/0.23/deploy/2025-05/spack/var/spack/environments/name` - the `2025-05` should be the only part that changes.

heatherkellyucl commented on spacksites/spack-env-templates/slurm/avx512.yaml:
```diff
@@ -0,0 +1,45 @@
+      - samtools@1.17
+    - - '%gcc@12.3.0'
+
+  include:
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_first.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu-on-gpu.yaml
+  - $HPC_SPACK_ROOT/spacksites/spack-env-templates/dev1/build/gpu_third.yaml
```
Should not be here.

heatherkellyucl commented on spacksites/spack-env-templates/slurm/gpu-on-gpu.yaml:
```diff
@@ -0,0 +1,34 @@
+  include_concrete:
+  - ${SPACK_ROOT}/var/spack/environments/gpu_first
+
+  modules:
+    default:
+      tcl:
+        projections:
+          # Named packages where we have more than one installed so need different module names.
+          # First match is picked, so more specific need to be at top. 'all' can be anywhere.
+          # Add omp for openmp variants.
+          fftw +openmp ^mpi+cuda: '{name}/{version}-{^mpi.name}-omp-cuda/{compiler.name}-{compiler.version}'
+          fftw +openmp ^mpi: '{name}/{version}-{^mpi.name}-omp/{compiler.name}-{compiler.version}'
+          fftw ~openmp ^mpi+cuda: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+          openblas threads=openmp: '{name}/{version}-omp/{compiler.name}-{compiler.version}'
+          plumed ^mpi+cuda: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+
+          ^cuda ^mpi: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+          ^mpi: '{name}/{version}-{^mpi.name}/{compiler.name}-{compiler.version}'
+          ^cuda: '{name}/{version}-cuda/{compiler.name}-{compiler.version}'
+          all:  '{name}/{version}/{compiler.name}-{compiler.version}'
+
+    prefix_inspections:
+      lib: [LD_LIBRARY_PATH, LIBRARY_PATH]
+      lib64: [LD_LIBRARY_PATH, LIBRARY_PATH]
+
```
Should go into `settings/initial_site_modules.yaml` - the end of it is there.

heatherkellyucl commented on spacksites/spack-env-templates/slurm/gpu_first.yaml:
```diff
@@ -0,0 +1,43 @@
+  include_concrete:
+  - ${SPACK_ROOT}/var/spack/environments/base
+
+  modules:
+    default:
+      tcl:
+        projections:
+          # Named packages where we have more than one installed so need different module names.
+          # First match is picked, so more specific need to be at top. 'all' can be anywhere.
+          # Add omp for openmp variants.
+          fftw +openmp ^mpi+cuda: '{name}/{version}-{^mpi.name}-omp-cuda/{compiler.name}-{compiler.version}'
+          fftw +openmp ^mpi: '{name}/{version}-{^mpi.name}-omp/{compiler.name}-{compiler.version}'
+          fftw ~openmp ^mpi+cuda: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+          openblas threads=openmp: '{name}/{version}-omp/{compiler.name}-{compiler.version}'
+          plumed ^mpi+cuda: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+
+          ^cuda ^mpi: '{name}/{version}-{^mpi.name}-cuda/{compiler.name}-{compiler.version}'
+          ^mpi: '{name}/{version}-{^mpi.name}/{compiler.name}-{compiler.version}'
+          ^cuda: '{name}/{version}-cuda/{compiler.name}-{compiler.version}'
+          all:  '{name}/{version}/{compiler.name}-{compiler.version}'
+
+    prefix_inspections:
+      lib: [LD_LIBRARY_PATH, LIBRARY_PATH]
+      lib64: [LD_LIBRARY_PATH, LIBRARY_PATH]
```
Same here, in `settings/initial_site_modules.yaml`

heatherkellyucl reviewed (changes requested):
Right, I had some pending comments, so those should show up now.

We are missing the equivalent of kathleen.yaml(`youngmichael.yaml` ?) that includes all these concrete environments. 

And from the now-outdated comment, multiarch,yaml should include
```
    - - intel-oneapi-mpi@2021.14.0 +external-libfabric
      # VASP is tested with 2023.2.0. +cluster gives scalapack
      - intel-oneapi-mkl@2023.2.0 +cluster mpi_family=mpich ^intel-oneapi-mpi
      - intel-oneapi-mkl@2023.2.0 +cluster mpi_family=openmpi ^openmpi%oneapi@2024.2.1
      - openmpi@4.1.6 fabrics=psm2,ucx,ofi schedulers=slurm +legacylaunchers
      - - '%oneapi@2024.2.1' # VASP is tested with 2024.2
  - intel-oneapi-compilers@2024.2.1
 ```

There does still need to be the `intel` builds in avx2 and avx512 that builds the end applications with the Intel stack.

```
      - vasp@6.5.1 +hdf5 ^intel-oneapi-mpi
      - hdf5@1.14.3 +fortran +mpi +hl
```

giordano committed:
Remove GPU environments from avx512 environment

giordano committed:
Move settings in GPU environments to global settings

giordano committed:
Add Intel packages

giordano committed:
Add settings for `castep`, `orca`, `vasp`

giordano committed:
Add concrete Young+Michael environment

giordano commented on spacksites/spack-env-templates/slurm/youngmichael.yaml:
```diff
@@ -0,0 +1,23 @@
+    unify: when_possible
+  view: false
+  include_concrete:
+  - /lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/environments/avx2
+  - /lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/environments/avx2_buildtools
+  - /lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/environments/avx512
+  - /lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/environments/multiarch
```
I don't think these paths exist yet (I saw that on Michael there's `/lustre/shared/ucl/apps/spack/0.23/`, but not a deployed system), so these are placeholders for the time being.

giordano reviewed (commented):
Ok, I hope I finally addressed all the comments

heatherkellyucl commented:
This was the comment I made above on the paths: 

"The paths for the Slurm environments should be `/lustre/apps/spack/0.23/deploy/2025-05/spack/var/spack/environments/name` - the `2025-05` should be the only part that changes."

It doesn't matter whether they exist yet or not on Young. The `/lustre/apps` paths are only going to exist on RHEL9 systems, backporting to RHEL7 will use `/lustre/shared/ucl/apps-by-arch/common/spack/0.23/deploy/` on Young and Michael. Which I can create.

giordano commented:
Ok, one issue I'm already facing is that we don't seem to have the oneapi compiler set up anywhere?  For example, the [first compiler environment](https://github.com/UCL-ARC/hpc-spack/blob/ff554330cc8a5bf77f63322df18cce43069ef317/spacksites/spack-env-templates/dev1/first_compiler.yaml) only builds gcc 12.3.0 (+ other build tools).  In the current code I see the oneapi compiler referenced by the environment at https://github.com/UCL-ARC/hpc-spack/blob/ff554330cc8a5bf77f63322df18cce43069ef317/spacksites/spack-env-templates/slurm/intel.yaml#L36 is that used somewhere?

giordano commented:
Also, compiling `charmpp` fails with
```
==> charmpp: Executing phase: 'install'
==> [2025-07-01-14:12:53.603475] './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j6' '--destination=/lustre/shared/ucl/apps-by-arch/common/spack/0.23/michael-young/spack/opt/spack/[padded-to-256-chars]/linux-rhel7-cascadelake/gcc-12.3.0/charmpp-6.10.2-3rgtiwf3svzq2n3mjf2k3yswv4g2dlie' 'openpmix' '--basedir=/lustre/shared/ucl/apps-by-arch/common/spack/0.23/michael-young/spack/opt/spack/[padded-to-256-chars]/linux-rhel7-cascadelake/gcc-12.3.0/pmix-5.0.3-raj6tzlx3nn5nfib5q7uk5jfffmemcnj' '--basedir=/lustre/shared/ucl/apps-by-arch/common/spack/0.23/michael-young/spack/opt/spack/[padded-to-256-chars]/linux-rhel7-cascadelake/gcc-12.3.0/ucx-1.17.0-kljnd7u3kc2gndcriuzil527fsph3jlv' 'smp' '--build-shared' '--with-production'
Error> option: openpmix is not supported in this version!
Supported compilers: clang craycc gcc icc iccstatic msvc pgcc xlc xlc64
Supported options: bigemulator bigsim common cuda flang gfortran ifort nolb omp ompipmix ooc papi pedantic perftools persistent pgf90 slurmpmi slurmpmi2 smp tsan
```

heatherkellyucl commented:
Ah yeah - charmpp will fail, so you may as well comment out namd.

Yes, we built VASP for the Slurm environments. What I did right now was this, but it does need adding properly somewhere:

```
spack -e base concretize -U -f
spack -e base install --only-concrete
spack compiler add $(spack location -i intel-oneapi-compilers)/compiler/latest/bin
# also need to add to environment: prepend_path: PATH: in linux/compilers.yaml
```
The last item in there is to add the equivalent of `/lustre/apps/spack/0.23/deploy/2025-05/spack/opt/spack/linux-rhel9-broadwell/gcc-12.3.0/intel-oneapi-compilers-2024.2.1-vrwg3ue262yl3wsaxl32qdhlwl7ypqfr/compiler/latest/bin` so `fpp` is found.

giordano committed:
Fix indentation

giordano committed:
Automatically detect top-dir paths in init spacksites scripts

Commit:
Git ignore Makefile

Commit:
`autotools` -> `automake`

Commit:
Change CUDA version, use absolute path of the concrete environment

giordano commented on spacksites/spack-env-templates/slurm/gpu_first.yaml:
```diff
@@ -0,0 +1,20 @@
+  specs:
+  # CUDA packages, compile them with gcc@12
+  - matrix:
+    - - 'cuda@12.2.2' # v12.2.2 matches the latest CUDA installed on Young
```
As mentioned on Slack, this is not necessarily the final version of `cuda` we'll use, I'm just mirroring the latest `cuda` module I found on current Young, but this is up for being changed.

giordano commented on spacksites/spack-env-templates/slurm/gpu_first.yaml:
```diff
@@ -0,0 +1,20 @@
+    unify: when_possible
+  view: false
+  include_concrete:
+  - /lustre/apps/spack/0.23/deploy/2025-02/spack/var/spack/environments/base
```
This is the absolute path of the environment I'm currently using, but I don't know if you want to retain this name or rather using a more meaningful one (I copied `2025-02` from https://github.com/UCL-ARC/hpc-spack/issues/64#issuecomment-2711378522 before realising I could have picked a different month......

giordano commented:
Compilation of `py-tensorflow@2.17.1` in the gpu_first environment failed with  
```
  >> 227    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:30: Error: no such instruction: `vmaxph %zmm7,%zmm0,%zmm0'
  >> 228    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:31: Error: no such instruction: `vmaxph %zmm6,%zmm1,%zmm1'
  >> 229    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:32: Error: no such instruction: `vmaxph %zmm5,%zmm2,%zmm2'
  >> 230    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:33: Error: no such instruction: `vmaxph %zmm4,%zmm3,%zmm3'
  >> 231    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:38: Error: no such instruction: `vmaxph %zmm1,%zmm0,%zmm0'
  >> 232    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:39: Error: no such instruction: `vmaxph %zmm3,%zmm2,%zmm2'
  >> 233    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:40: Error: no such instruction: `vmaxph %zmm2,%zmm0,%zmm0'
  >> 234    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:50: Error: no such instruction: `vmaxph %zmm1,%zmm0,%zmm0'
  >> 235    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:62: Error: no such instruction: `vmaxph %ymm1,%ymm0,%ymm0'
  >> 236    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:64: Error: no such instruction: `vmaxph %xmm1,%xmm0,%xmm0'
  >> 237    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:66: Error: no such instruction: `vmaxph %xmm1,%xmm0,%xmm0'
  >> 238    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:68: Error: no such instruction: `vmaxph %xmm1,%xmm0,%xmm0'
  >> 239    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:70: Error: no such instruction: `vmaxsh %xmm1,%xmm0,%xmm0'
  >> 240    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:90: Error: no such instruction: `vmaxph %zmm1,%zmm0,%zmm0{%k1}'
  >> 241    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:126: Error: no such instruction: `vminph %zmm5,%zmm1,%zmm1'
  >> 242    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:127: Error: no such instruction: `vmaxph %zmm5,%zmm0,%zmm0'
  >> 243    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:128: Error: no such instruction: `vminph %zmm4,%zmm11,%zmm11'
  >> 244    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:129: Error: no such instruction: `vmaxph %zmm4,%zmm10,%zmm10'
  >> 245    /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccVthKJn.s:130: Error: no such instruction: `vminph %zmm3,%zmm7,%zmm7'
```
I need to figure out _why_, my hunch is that an old `as` is being used.

giordano commented:
Confirmed that GCC is picking up system's `as` instead of the one from binutils built as part of the environment:
```console
$ spack -e gpu_first build-env py-tensorflow -- bash
$ echo '' | gcc -v -x c - -o /dev/null
Reading specs from /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/specs
COLLECT_GCC=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/bin/gcc
COLLECT_LTO_WRAPPER=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /lustre/apps/spack/0.23/deploy/2025-02/build_stage/ccspapp/spack-stage-gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/spack-src/configure --prefix=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv --with-pkgversion='Spack GCC' --with-bugurl=https://github.com/spack/spack/issues --disable-multilib --enable-languages=c,c++,fortran --disable-nls --disable-canonical-system-headers --with-system-zlib --with-zstd-include=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zstd-1.5.5-h4aqnii7gc7sli5ywxcsfvsd4jzfq5ee/include --with-zstd-lib=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zstd-1.5.5-h4aqnii7gc7sli5ywxcsfvsd4jzfq5ee/lib --enable-bootstrap --with-mpfr-include=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpfr-4.2.0-toffeyegi5q7zyisklxdp6ske2vp33q2/include --with-mpfr-lib=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpfr-4.2.0-toffeyegi5q7zyisklxdp6ske2vp33q2/lib --with-gmp-include=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gmp-6.2.1-74hwjamabmd5oolnxdf2sybl326ydvpv/include --with-gmp-lib=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gmp-6.2.1-74hwjamabmd5oolnxdf2sybl326ydvpv/lib --with-mpc-include=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpc-1.3.1-3vbb72iltfgzbjiwqwlkm4stjblczaop/include --with-mpc-lib=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpc-1.3.1-3vbb72iltfgzbjiwqwlkm4stjblczaop/lib --without-isl --with-stage1-ldflags='-Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib64 -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpfr-4.2.0-toffeyegi5q7zyisklxdp6ske2vp33q2/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zlib-1.3.1-baqgl5hqvqlcxph4dgwoexnzxgidtlaf/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpc-1.3.1-3vbb72iltfgzbjiwqwlkm4stjblczaop/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gmp-6.2.1-74hwjamabmd5oolnxdf2sybl326ydvpv/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zstd-1.5.5-h4aqnii7gc7sli5ywxcsfvsd4jzfq5ee/lib -Wl,-rpath,/usr/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-runtime-11.5.0-v3baia74xkligd35kzwuzlqdy7dxyogy/lib -Wl,-rpath,/usr/lib64' --with-boot-ldflags='-Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib64 -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpfr-4.2.0-toffeyegi5q7zyisklxdp6ske2vp33q2/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zlib-1.3.1-baqgl5hqvqlcxph4dgwoexnzxgidtlaf/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/mpc-1.3.1-3vbb72iltfgzbjiwqwlkm4stjblczaop/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gmp-6.2.1-74hwjamabmd5oolnxdf2sybl326ydvpv/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/zstd-1.5.5-h4aqnii7gc7sli5ywxcsfvsd4jzfq5ee/lib -Wl,-rpath,/usr/lib -Wl,-rpath,/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-runtime-11.5.0-v3baia74xkligd35kzwuzlqdy7dxyogy/lib -Wl,-rpath,/usr/lib64 -static-libstdc++ -static-libgcc' --with-build-config=spack
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.3.0 (Spack GCC)
COLLECT_GCC_OPTIONS='-v' '-o' '/dev/null' '-mtune=generic' '-march=x86-64' '-dumpdir' 'a-'
 /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/cc1 -quiet -v - -quiet -dumpdir a- -dumpbase - -mtune=generic-march=x86-64 -version -o /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccyr9eEw.s
GNU C17 (Spack GCC) version 12.3.0 (x86_64-pc-linux-gnu)
        compiled by GNU C version 12.3.0, GMP version 6.2.1, MPFR version 4.2.0, MPC version 1.3.1, isl version none
GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../../../x86_64-pc-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/include
 /usr/local/include
 /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/include
 /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/include-fixed
 /usr/include
End of search list.
GNU C17 (Spack GCC) version 12.3.0 (x86_64-pc-linux-gnu)
        compiled by GNU C version 12.3.0, GMP version 6.2.1, MPFR version 4.2.0, MPC version 1.3.1, isl version none
GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
Compiler executable checksum: 2dceacfbc8ff80ccd41a1a9e7d22ac14
COLLECT_GCC_OPTIONS='-v' '-o' '/dev/null' '-mtune=generic' '-march=x86-64' '-dumpdir' 'a-'
 as -v --64 -o /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccPHcGQD.o /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccyr9eEw.s
GNU assembler version 2.35.2 (x86_64-redhat-linux) using BFD version version 2.35.2-63.el9
COMPILER_PATH=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/
LIBRARY_PATH=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/dev/null' '-mtune=generic' '-march=x86-64' '-dumpdir' 'a.' /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/collect2 -plugin /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/liblto_plugin.so -plugin-opt=/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/libexec/gcc/x86_64-pc-linux-gnu/12.3.0/lto-wrapper -plugin-opt=-fresolution=/lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/cctVDbp9.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o /dev/null /lib/../lib64/crt1.o /lib/../lib64/crti.o /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/crtbegin.o -L/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0 -L/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../.. -rpath /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib64 /lustre/apps/spack/0.23/deploy/2025-02/spack/../build_stage/ccPHcGQD.o -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /lustre/apps/spack/0.23/deploy/2025-02/spack/opt/spack/linux-rhel9-cascadelake/gcc-11.5.0/gcc-12.3.0-ut7qt62bwv4rmty4j5pjbfxsisawlekv/lib/gcc/x86_64-pc-linux-gnu/12.3.0/crtend.o /lib/../lib64/crtn.o
/usr/bin/ld: /lib/../lib64/crt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status
$ which as
/usr/bin/as
$ /usr/bin/as --version
GNU assembler version 2.35.2-63.el9
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or later.
This program has absolutely no warranty.
This assembler was configured for a target of `x86_64-redhat-linux'.
```

My understanding is that we need to rebuild gcc with the `+binutils` option, which forces Spack's gcc to use its own binutils tools instead of relying on the system ones.  In Spack they even have these checks for compatibility of GCC + system binutils on various Red Hat versions, but not for RHEL 9: https://github.com/spack/spack-packages/blob/10cb8833e524faca0ed0761f7e426311436724d8/repos/spack_repo/builtin/packages/gcc/package.py#L414-L424

***Edit***: confirmed that the issue above is fixed by compiling gcc together with binutils: `gcc@12.3.0 +binutils`.

Commit:
Build `gcc` with `+binutils` flag

Without it, GCC will rely on system tools (e.g. assembler and linker) which may
be too old for compiling specific code.

giordano commented on spacksites/spack-env-templates/slurm/avx2_buildtools.yaml:
```diff
@@ -0,0 +1,23 @@
+      - perl@5.38.0
+      - texinfo
+    - - '%gcc@12.3.0'
+  - gcc@12.3.0 +binutils
```
@heatherkellyucl do we need this gcc? It's different from the first gcc compiler (which is compiled with system gcc), used to build all packages (including this one), so we end up with two different gccs in the environment.

heatherkellyucl replied:
@giordano The compiler module in use ought to be `compilers/gcc/12.3.0/gcc-12.3.0` which on Kathleen's deployed stack is /lustre/apps/spack/0.23/deploy/2025-05/spack/opt/spack/linux-rhel9-cascadelake/gcc-12.3.0/gcc-12.3.0-eent72vfu2pwphhi42l345ropmw3tb2t/

For some of the environments, having it listed like that was to make sure it was a root spec in the environment and got a module.

giordano replied:
That's not what I'm seeing.  When I create and concretize `base` with
```
spack env create base spacksites/spack-env-templates/slurm/base.yaml
spack -e base concretize -U -f
```
I see the following spec
```
 -   qsg4r3d  gcc@12.3.0%gcc@12.3.0+binutils+bootstrap~graphite~mold~nvptx~piclibs~profiled~strip build_system=autotools build_type=RelWithDebInfo languages='c,c++,fortran' arch=linux-rhel9-cascadelake                                   
```
but this is a new gcc, (note the leading `-` instead of `[e]`, and the compiler is `%gcc@12.3.0`, not `%gcc@11.5.0` which would be the system one) not the one already compiled.

heatherkellyucl replied:
Yes, that's what I meant: the base environment should be using a gcc built with gcc 12 (which ends up as `compilers/gcc/12.3.0/gcc-12.3.0`). And everything else should be using that one.

giordano replied:
Do we need to add this `gcc` to the compilers then?  In https://github.com/UCL-ARC/hpc-spack/issues/64#issuecomment-2711378522 I only see
```
spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/bin
```
for the Intel oneAPI compiler, but not something similar for gcc.

heatherkellyucl replied:
Ah right, I see. We aren't doing that and yes, we ought to be.

giordano replied:
Spack doesn't seem to like the duplicate compiler now though:
```console
$ spack compiler add $(spack location -i gcc@12.3.0%gcc@12.3.0)
==> Added 1 new compiler to /lustre/apps/spack/0.23/deploy/2025-07/spack/etc/spack/linux/compilers.yaml
    gcc@12.3.0
==> Compilers are defined in the following files:
    /lustre/apps/spack/0.23/deploy/2025-07/spack/etc/spack/compilers.yaml  /lustre/apps/spack/0.23/deploy/2025-07/spack/etc/spack/linux/compilers.yaml
$ spack env create avx512 spacksites/spack-env-templates/slurm/avx512.yaml
==> Created environment avx512 in: /lustre/apps/spack/0.23/deploy/2025-07/spack/var/spack/environments/avx512
==> Activate with: spack env activate avx512
$ spack -e avx512 concretize -U -f
==> Warning: duplicate found for gcc@=12.3.0 on rhel9/x86_64. Edit your compilers.yaml configuration to remove it.
==> Warning: using "pmix@3.2.3" which is a deprecated version
==> Warning: duplicate found for gcc@=12.3.0 on rhel9/x86_64. Edit your compilers.yaml configuration to remove it.
==> Warning: duplicate found for gcc@=12.3.0 on rhel9/x86_64. Edit your compilers.yaml configuration to remove it.
[...]
```

giordano committed:
Update some specs and paths

giordano committed:
Compile GPU environments for avx2 (`x86_64_v3`)

giordano committed:
Do not include `base` environment in GPU ones

They have packages built for incompatible microarchitectures anyway, not much
value in including it, only pain.

giordano committed:
Use requirements for target specification in avx2 environments

Requirements are higher-priority than reuse, which would otherwise mix up avx512
packages in the avx2 environment, wreaking havoc, see
<https://spack.readthedocs.io/en/latest/frequently_asked_questions.html#why-does-spack-pick-particular-versions-and-variants>.

giordano committed:
Require avx2 target also for first compiler

Otherwise gcc and related libraries (most notable libstdc++) will be unusable on
avx2 nodes, leading to frequent SIGILL errors.

giordano commented:
This is turning into an endless rabbit hole.  Current situation is:

* we have two GCCs, one built for `cannonlake` (ISA of login node) and one built for `x86_64_v3` (~AVX2, good for GPU nodes)
  * this is problematic because Spack can't really distinguish them (in the list of compilers, the fact that a compiler was built for a specific microarchitecture isn't recorded anywhere) and when resolving environments it will pick one of them in an unspecified.  Relatedly, resolving all environments always results in a warning about duplicate `gcc` (because, again, Spack can't tell them apart). I'm constantly getting the `cannonlake` build being used for compiling everywhere, also when targeting `x86_64_v3`, and then C++ libraries/executables link to the avx512 `libstdc++` from `gcc-runtime`, which leads to SIGILL upon startup when executing them on avx2 nodes
  * we should _really_ have a single GCC, built only targeting `x86_64_v3`, so that it's usable everywhere. `libstdc++` is a runtime library used for stuff like throwing errors, nothing exceptionally performance critical
* by manually forcing the compiler to be the `x86_64_v3` one, I made some progress, but now installing the `gpu_first` environment on the GPU node fails when compiling `py-tensorstore` (I think it's a dependency of alphafold) with
   ```
     >> 153    bazel-out/k8-opt/bin/external/net_sourceforge_half/_virtual_includes/half/half.hpp:911:62: error: '_MM_FROUND_TO_NEAREST_INT' was not declared in this scope
        154        911 |                                 (R==std::round_to_nearest) ? _MM_FROUND_TO_NEAREST_INT :
        155            |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
     >> 156      bazel-out/k8-opt/bin/external/net_sourceforge_half/_virtual_includes/half/half.hpp:912:63: error: '_MM_FROUND_TO_ZERO' was not declared in this scope
        157        912 |                                 (R==std::round_toward_zero) ? _MM_FROUND_TO_ZERO :
        158            |                                                               ^~~~~~~~~~~~~~~~~~
   ```
   `_MM_FROUND_TO_NEAREST_INT` and `_MM_FROUND_TO_ZERO` are SSE4.1 macros, but you need to include `immintrin.h` to have them available.  Apparently the `half` package [forgot to do this properly](https://sourceforge.net/p/half/code/415/) (perhaps older GCC version happened to expose the macros when including other header files, they have done similar things in the past).  This was noted also in another package vendoring `half`: https://github.com/RenderKit/openvkl/issues/12.  So now I need to figure out how to patch up `py-tensorstore` to use the fixed `half` (and of course `py-tensorstore` is a Google project which uses Bazel, which makes everything miserable)

giordano committed:
Add `py-tensorstore` package to fix upstream bug in vendored `half` dependency

giordano committed:
Move all compilers to `avx2_buildtools`

There were two related problems:

* we had two GCCs, which caused confusion to Spack (it sees them as duplicates
  and can't distinguish them)
* if Spack picks up the avx512 build of gcc, C++ binaries are buing linked to
  avx512 `libstdc++`, which triggers SIGILL errors upon loading when run on avx2
  nodes.

The solution to above issues is to remove the avx512 builds of the compilers
from the mix and only keep the avx2 ones, and avoid any confusion and picking up
wrong compilers.

giordano committed:
Update again paths of concrete environments

giordano committed:
Use only `cuda_arch=80` for CP2K

giordano committed:
Update paths of included environments

giordano committed:
Make stronger requirement for avx2 target in multiarch environment

giordano committed:
Remove yet another GCC

giordano commented:
Ok, I believe this is mostly in a working state.  This is more or less the list of commands I ran (adapted from https://github.com/UCL-ARC/hpc-spack/issues/64#issuecomment-2711378522):
```sh
# RHEL9
source spacksites/myriad-utilities/init-spacksites-rhel9.sh
site_name=2025-08
spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini create ${site_name}

# trust existing buildcache PGP keys without having to run below eval first
spacksites/process-env-scripts/spack-deps-spack-env-run.sh /lustre/apps/spack/0.23/deploy/${site_name}/spack/share/spack/setup-env.sh spack gpg trust /lustre/apps/spack/0.23/buildcache/build_cache/_pgp/19FBA61A794B66335BA0E28AB00983609A7C5CC4.pub
spacksites/process-env-scripts/spack-deps-spack-env-run.sh /lustre/apps/spack/0.23/deploy/${site_name}/spack/share/spack/setup-env.sh spack gpg trust /lustre/apps/spack/0.23/buildcache/build_cache/_pgp/340EE729AF82B40318C2ADA269B9BEA0F177B630.pub
spacksites/process-env-scripts/spack-deps-spack-env-run.sh /lustre/apps/spack/0.23/deploy/${site_name}/spack/share/spack/setup-env.sh spack gpg trust /lustre/apps/spack/0.23/buildcache/build_cache/_pgp/5E39DBFE9E57AE568737C00DC4039802F0B4E6E3.pub

# build first compiler
spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini install-env ${site_name} compiler first_compiler.yaml
eval $(spacksites/spacksites --config-file spacksites/settings/deploy_sites.ini spack-setup-env ${site_name})

# update buildcache
spack -e compiler buildcache push /lustre/apps/spack/0.23/buildcache
spack buildcache update-index /lustre/apps/spack/0.23/buildcache

# the -U -f are only really needed when the environment already exists
spack env create avx2_buildtools spacksites/spack-env-templates/slurm/avx2_buildtools.yaml
spack -e avx2_buildtools concretize -U -f
spack -e avx2_buildtools env depfile -o Makefile
make --output-sync=target -j40
spack -e avx2_buildtools buildcache push /lustre/apps/spack/0.23/buildcache

spack compiler add $(spack -e avx2_buildtools location -i gcc@12.3.0%gcc@12.3.0)
spack compiler add $(spack -e avx2_buildtools location -i intel-oneapi-compilers)/compiler/latest/bin

spack env create base spacksites/spack-env-templates/slurm/base.yaml
spack -e base concretize -U -f
spack -e base env depfile -o Makefile
make --output-sync=target -j40
spack -e base buildcache push /lustre/apps/spack/0.23/buildcache

spack env create avx512 spacksites/spack-env-templates/slurm/avx512.yaml
spack -e avx512 concretize -U -f
spack -e avx512 env depfile -o Makefile
make --output-sync=target -j40
spack -e avx512 buildcache push /lustre/apps/spack/0.23/buildcache

spack env create gpu_first spacksites/spack-env-templates/slurm/gpu_first.yaml
spack -e gpu_first concretize -U -f
spack -e gpu_first env depfile -o Makefile
make --output-sync=target -j40
spack -e gpu_first buildcache push /lustre/apps/spack/0.23/buildcache

spack env create gpu-on-gpu spacksites/spack-env-templates/slurm/gpu-on-gpu.yaml
spack -e gpu-on-gpu concretize -U -f
spack -e gpu-on-gpu env depfile -o Makefile
make --output-sync=target -j40
spack -e gpu-on-gpu buildcache push /lustre/apps/spack/0.23/buildcache

spack env create gpu_third spacksites/spack-env-templates/slurm/gpu_third.yaml
spack -e gpu_third concretize -U -f
spack -e gpu_third env depfile -o Makefile
make --output-sync=target -j40
spack -e gpu_third buildcache push /lustre/apps/spack/0.23/buildcache

spack env create avx2 spacksites/spack-env-templates/slurm/avx2.yaml
spack -e avx2 concretize -U -f
spack -e avx2 env depfile -o Makefile
make --output-sync=target -j40
spack -e avx2 buildcache push /lustre/apps/spack/0.23/buildcache

spack env create multiarch spacksites/spack-env-templates/slurm/multiarch.yaml
spack -e multiarch concretize -U -f
spack -e multiarch env depfile -o Makefile
make --output-sync=target -j40
spack -e multiarch buildcache push /lustre/apps/spack/0.23/buildcache
```
I said "more or less" because in practice I didn't do exactly that linearly from scratch, due to the presence of the buildcache, but this is what I believe is the right order of the steps, if we were to restart from scratch.

For the record, I built the three GPU environments on the GPU node, the rest all on the login node, but the only reason for which I compiled `gpu_first` on the GPU node was that `py-jaxlib` (dependency of `py-alphafold`) was being forcibly compiled for avx512 despite the request to target `x86_64_v3` (I believe bazel was determining the list of files to compile based on the host, rather than looking at the capabilities of the compiler which is what other more sensible build systems would do), I think just moving `py-alphafold` to the `gpu-on-gpu` environment would address that issue.

Note: I'm getting a compilation error in the `multiarch` environment while compiling `intel-tbb`:
```
115    [ 88%] Linking CXX shared library ../../gnu_12.3_cxx11_64_release/libtbbbind_2_0.so
     116    cd /lustre/apps/spack/0.23/deploy/2025-08/build_stage/ccspapp/spack-stage-intel-tbb-2021.12.0-vgw5yvotqxnmlpxzi3plgxb5iz4abeeo/spack-build-vgw5yvo/src/tbbbind && /lustre/apps/spack/0.23/deplo
            y/2025-08/spack/opt/spack/linux-rhel9-x86_64_v3/gcc-12.3.0/cmake-3.30.5-vrg6hezjrjh7ivtgj5npzrkw3f425qib/bin/cmake -E cmake_link_script CMakeFiles/tbbbind_2_0.dir/link.txt --verbose=1
     117    /lustre/apps/spack/0.23/deploy/2025-08/spack/lib/spack/env/gcc/g++ -fPIC -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -Wl,--version-script="/lustre/apps/spack/0.23/deploy/2025-08/build_stage/
            ccspapp/spack-stage-intel-tbb-2021.12.0-vgw5yvotqxnmlpxzi3plgxb5iz4abeeo/spack-src/src/tbbbind/def/lin64-tbbbind.def" -Wl,-z,relro,-z,now,-z,noexecstack -shared -Wl,-soname,libtbbbind_2_0.so.
            3 -o ../../gnu_12.3_cxx11_64_release/libtbbbind_2_0.so.3.12 CMakeFiles/tbbbind_2_0.dir/tbb_bind.cpp.o  -Wl,-rpath,:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
            :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
            ::::::::::::::::::::::: /usr/lib64/libhwloc.so -ldl
  >> 118    /lustre/apps/spack/0.23/deploy/2025-08/spack/opt/spack/linux-rhel9-x86_64_v3/gcc-12.3.0/binutils-2.43.1-2tgfs7s3be5tle6fx7v6c2ormkfdfuat/bin/ld: cannot open output file ../../gnu_12.3_cxx11_6
            4_release/libtbbbind_2_0.so.3.12: No such file or directory
  >> 119    collect2: error: ld returned 1 exit status
  >> 120    make[3]: *** [src/tbbbind/CMakeFiles/tbbbind_2_0.dir/build.make:102: gnu_12.3_cxx11_64_release/libtbbbind_2_0.so.3.12] Error 1
     121    make[3]: Leaving directory '/lustre/apps/spack/0.23/deploy/2025-08/build_stage/ccspapp/spack-stage-intel-tbb-2021.12.0-vgw5yvotqxnmlpxzi3plgxb5iz4abeeo/spack-build-vgw5yvo'
  >> 122    make[2]: *** [CMakeFiles/Makefile2:251: src/tbbbind/CMakeFiles/tbbbind_2_0.dir/all] Error 2
```
I don't have a good intuition of what may be going on here 😞 ***Edit***: it was a race condition in the build system, compiling with fewer parallel jobs was successful 🤦
